### PR TITLE
Add LazyList

### LINKED ISSUE #1395: Experiment with LazyList
Tasks

- API surface (i.e. collect information from other languages about what operations are usually available).

### PR DESCRIPTION
**Pull request by esbenbjerre**
- [x] Implement LazyList with strict head
- [x] Implement LazyList2 which is essentially a Lazy[LazyList] (maybe this should be an opaque type?)
- [x] Add documentation
- [x] Ensure that we have the same set of base functions
- [x] Add comment to documentation if function forces the list
- [x] Add tests
- [x] Brainstorm 2-3 programs that use Lazy and would be suitable for flix.dev

Fixes #1395.

Defined functions:
- empty
- repeat
- from
- isEmpty
- length
- head
- last
- append
- reverse
- take
- drop
- tail
- toList
- fold
- foldLeft
- foldRight
- exists
- forall
- map
- filter
- unfold
- zipWith
- zip
- __eq

### CODE REVIEW

File: main/src/library/LazyList.flix [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    /// An immutable lazy list.
+    ///
+    /// A lazy list is either the empty list represented by `Empty`, or
+    /// an element `x` followed by a lazy list `xs`.
+    /// Note that both the head and tail of a lazy list is lazy.
+    ///
+    pub enum LazyList[t] {
+        case Empty,
+        case Cons(Lazy[t], Lazy[LazyList[t]])
+    }
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:29]: There are actually two types of lazy lists that I would like to consider, but neither is this one :)

The first is:

```
    pub enum LazyList[t] {
        case Empty,
        case Cons(t, Lazy[LazyList[t]])
    }
```

But why is the element `t` not lazy you ask? It turns out to be unnecessary, since I can do `List[Lazy[Int]]` or `LazyList[Lazy[Int]]`. The second definition -- lets call it LazyList2 for now -- captures the intuition that you want to have a list where not even the first element is evaluated:

```
    pub enum LazyList2[t] {  
      case LazyList2(Lazy[LList[t]])
    }
    pub enum InnerLazyList[t] {
        case Empty,
        case Cons(t, Lazy[InnerLazyList[t]])
    }
```

Comment (magnus-madsen) [reply main/src/library/LazyList.flix:29]: Does that make sense? :)

File: main/src/library/LazyList.flix [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    pub def fold(f: (b, a) -> b & e, s: b, xs: LazyList[a]): b & e = foldLeft(f, s, xs)
+
+    pub def foldLeft(f: (b, a) -> b & e, s: b, xs: LazyList[a]): b & e = match xs {
+        case Empty => s
+        case Cons(h, t) => foldLeft(f, f(s, force h), force t)
+    }
+
+    /// Redundancy Error: Unconditionally recursive definition 'constant'. All branches will recurse indefinitely.
+    /// Well yes, but actually no...
+    pub def constant(x: a): LazyList[a] =
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:50]: This might still be a problem, if so, open a ticket and we can discuss :)

File: main/src/library/LazyList.flix [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        foldRight((x, acc) -> cons(x, acc), ys, xs)
+
+    pub def fold(f: (b, a) -> b & e, s: b, xs: LazyList[a]): b & e = foldLeft(f, s, xs)
+
+    pub def foldLeft(f: (b, a) -> b & e, s: b, xs: LazyList[a]): b & e = match xs {
+        case Empty => s
+        case Cons(h, t) => foldLeft(f, f(s, h), force t)
+    }
+
+    pub def foldRight(f: (a, b) -> b & e, s: b, xs: LazyList[a]): b & e = match xs {
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:65]: Question: Will this not force the entire list? (and hence e.g. exists would force the entire list, which should be unnecessary?)

Comment (esbenbjerre) [reply main/src/library/LazyList.flix:65]: I think you're right. I took inspiration from Functional Programming in Scala, but I missed that they do the 2nd argument to `f` by-name.

**Comment by magnus-madsen** [comment]
Also, for our research purposes, I would actually like if you could implement each function separately, without relying too much on foldLeft/foldRight. This will be helpful later (for me).

**Comment by esbenbjerre** [comment]
> Also, for our research purposes, I would actually like if you could implement each function separately, without relying too much on foldLeft/foldRight. This will be helpful later (for me).

Roger :-)

**Comment by magnus-madsen** [comment]
Let me know when this is ready for review :)

File: main/src/library/LazyList.flix (hunk 3 above)
Comment (esbenbjerre) [reply main/src/library/LazyList.flix:65]: I'm a bit stuck with this method. To prevent forcing the entire list it is possible to define it as seen below. But I do not think this is the type signature we want.
```
pub def foldRight(f: (a, () -> b) -> b, s: () -> b, xs: LazyList[a]): b = match xs {
    case Empty => s()
    case Cons(h, t) => f(h, () -> foldRight(f, s, force t))
}
```

Comment (esbenbjerre) [reply main/src/library/LazyList.flix:65]: After thinking some more I'm actually not sure the above works.

**Comment by magnus-madsen** [comment]
I believe @JonathanStarup is almost done with backend support for `lazy`- so hopefully soon this can be tested.

**Comment by esbenbjerre** [comment]
> I believe @JonathanStarup is almost done with backend support for `lazy`- so hopefully soon this can be tested.

Sounds good! I'll try to wrap it up ASAP.

File: main/src/library/LazyList.flix (hunk 3 above)
Comment (magnus-madsen) [reply main/src/library/LazyList.flix:65]: I think *I* was confused. This stuff is also new to me.

A fold right of `1 :: 2 :: 3 :: Nil` looks like `f(1, f(2, f(3, ...)))`. This means, the first thing to do is to apply `f` to the last element of the list. This requires the entire list to be forced.

Hmm but maybe not:

http://voidmainargs.blogspot.com/2011/08/folding-stream-with-scala.html

Now I am confused on a higher level.

Comment (magnus-madsen) [reply main/src/library/LazyList.flix:65]: https://stackoverflow.com/questions/48654292/why-does-foldright-work-for-infinite-lists

**Comment by magnus-madsen** [comment]
> > I believe @JonathanStarup is almost done with backend support for `lazy`- so hopefully soon this can be tested.
> 
> Sounds good! I'll try to wrap it up ASAP.

Not to worry. I think there will need to be some experimentation done (see above w.r.t. foldRight).
(If you need another task to work on, there is always #1448)

**Comment by magnus-madsen** [comment]
I would like to know if the equivalent of this:

```scala
Stream.continually(true).foldRight(false)(_ || _)
```

works for either list?

**Comment by magnus-madsen** [comment]
This seems interesting: http://voidmainargs.blogspot.com/2011/08/folding-stream-with-scala.html

**Comment by esbenbjerre** [comment]
> I would like to know if the equivalent of this:
> 
> ```scala
> Stream.continually(true).foldRight(false)(_ || _)
> ```
> 
> works for either list?

It does not.



> This seems interesting: http://voidmainargs.blogspot.com/2011/08/folding-stream-with-scala.html
I've come across upon this article previously. However, I don't think we can implement it like that as of now.
The thing is that Scala has fancy sugar for by-name parameters, e.g. `f(x: => Boolean)`. All solutions I've seen so far uses by-name parameters (including the one above) such that the "real" type signature is hidden. As I understand it, `f(x: => Boolean)` is desugared into `f(x: () => Boolean)`, and each use of `x` will be desugared into `x()` when needed. The following works in Scala.
```scala
def foldr[A, B](f: (A, => B) => B, s: B, xs: LazyList[A]): B = xs match {
  case Empty => s
  case Cons(head, tail) => f(head, foldr(f, s, tail()))
}
```
But changing the signature to `def foldr[A, B](f: (A, B) => B, s: B, xs: LazyList[A]): B` will cause stack overflow for something like `Stream.continually(true).foldRight(false)(_ || _)`.
TL;DR I haven't figured out how to make it work without having the second argument of `f` be a thunk.

**Comment by magnus-madsen** [comment]
> will cause stack overflow for something like `Stream.continually(true).foldRight(false)(_ || _)`.
> TL;DR I haven't figured out how to make it work without having the second argument of `f` be a thunk.

I think the  second argument has to be a thunk `Lazy[b]`.

**Comment by magnus-madsen** [comment]
I have merged support for `lazy/force` to the backend.

File: main/src/library/LazyList.flix [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+    pub def zip(xs: LazyList[a], ys: LazyList[b]): LazyList[(a, b)] = zipWith((x, y) -> (x, y), xs, ys)
+
+    pub def __eq(xs: LazyList[a], ys: LazyList[a]): Bool = match (xs, ys) {
+        case (Empty, Empty) => true
+        case (Cons(h1, t1), Cons(h2, t2)) => (h1 == h2) && __eq(force t1, force t2)
+        case _ => false
+    }
+
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:110]: A few more operations I'd like to see:
- drop
- take
- filter
- head
- last (!)
- reverse
- unfold/unfoldWithiter/unfoldWithOk

**Comment by esbenbjerre** [comment]
> I think the second argument has to be a thunk `Lazy[b]`.

As in `b: () -> Lazy[b]`?

**Comment by magnus-madsen** [comment]
> 
> 
> > I think the second argument has to be a thunk `Lazy[b]`.
> 
> As in `b: () -> Lazy[b]`?

```
def foldr[A, B](f: (A, Lazy[B]) => B, s: B, xs: LazyList[A]): B
```

I think. Not sure about the return type.

**Comment by magnus-madsen** [comment]
It would be nice, when you write tests, that we can ensure that the list is not evaluated more than necessary. For example:

```
head(1 `cons` ???)
```

should not blow up.

**Comment by esbenbjerre** [comment]
I'm close to a good implementation now. Even the "regular" LazyList is fun to play with, we can do
```
def main(): Option[Int32] =
    LazyList.from(1) |>
    LazyList.map(x -> x * x) |>
    LazyList.filter(x -> x > 16) |>
    LazyList.head
```
where the initial list is the infinite list [1, ...]

**Comment by magnus-madsen** [comment]
I am also considering whether I should add `#::`, as in `1 #:: 2 #:: LazyNil` as a short hand for cons. But maybe this is something we should do a bit later once we have experimented more.

File: main/src/library/LazyList.flix (hunk 4 above)
Comment (esbenbjerre) [reply main/src/library/LazyList.flix:110]: I think I'm currently unable to implement `unfoldWithIter` and `unfoldWithOkIter` for LazyList2 as they are impure.
```
>> Unable to unify the Boolean formulas: 'true' and 'false'.

122 |     pub def unfoldWithIter(next: () ~> Option[a]): LazyList2[a] & Impure = LazyList2(lazy LazyList.unfoldWithIter(next))
                                                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                                                                           mismatched boolean formulas.
```

Comment (magnus-madsen) [reply main/src/library/LazyList.flix:146]: Ah yes, I guess we have to leave those for now.

**Comment by esbenbjerre** [comment]
Ideas for lazy programs.

1. An infinite sequence of prime numbers (requires that we add `def mod(x: Int32, n: Int32): Int32 = x - n * (x / n)` to Int32).
```
use LazyList.LazyList;

def isPrime(p: Int32): Bool =
    LazyList.from(2) |>
    LazyList.take(p - 2) |>
    LazyList.forall(x -> Int32.mod(p, x) != 0)

def primes(): LazyList[Int32] =
    LazyList.from(2) |>
    LazyList.filter(isPrime)
```

**Comment by magnus-madsen** [comment]
> 
> 
> Ideas for lazy programs.
> 
>     1. An infinite sequence of prime numbers (requires that we add `def mod(x: Int32, n: Int32): Int32 = x - n * (x / n)` to Int32).
> 
> 
> ```
> use LazyList.LazyList;
> 
> def isPrime(p: Int32): Bool =
>     LazyList.from(2) |>
>     LazyList.take(p - 2) |>
>     LazyList.forall(x -> Int32.mod(p, x) != 0)
> 
> def primes(): LazyList[Int32] =
>     LazyList.from(2) |>
>     LazyList.filter(isPrime)
> ```

Looks good. I wonder if we can also do https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes or Fibonacci?

File: main/src/library/LazyList.flix [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    pub def repeat(x: a): LazyList[a] = LazyCons(x, lazy repeat(x))
+
+    ///
+    /// Returns an infinite sequence of integers starting from and including `n`.
+    ///
+    pub def from(n: Int32): LazyList[Int32] = LazyCons(n, lazy from(n + 1))
+
+    ///
+    /// Returns true if and only if `xs` is the empty LazyList, i.e. `Empty`.
+    ///
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/LazyList.flix:49]: Would you like me to add a comment that explicitly says this also?

File: main/src/library/LazyList.flix [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+
+    ///
+    /// Returns an infinite LazyList of `x`.
+    ///
+    pub def repeat(x: a): LazyList[a] = LazyCons(x, lazy repeat(x))
+
+    ///
+    /// Returns an infinite sequence of integers starting from and including `n`.
+    ///
+    pub def from(n: Int32): LazyList[Int32] = LazyCons(n, lazy from(n + 1))
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/LazyList.flix:45]: Yes that is the idea. Your example would run forever since from is infinite, but something like `LazyList.length(LazyList.take(100000, LazyList.from(0)))` returns 100000 without overflowing the stack.

File: main/src/library/LazyList.flix [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+            else if (n == 1) LazyCons(h, lazy Empty)
+            else Empty
+    }
+
+    ///
+    /// Returns `xs` without the first `n` elements.
+    ///
+    pub def drop(n: Int32, xs: LazyList[a]): LazyList[a] = match xs {
+        case Empty => xs
+        case LazyCons(_, t) => if (n > 0) drop(n - 1, force t) else xs
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/LazyList.flix:118]: I agree. I have no idea why I did it this way, I would normally use the style you suggest.

File: main/src/library/LazyList.flix [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        case Empty => s
+        case LazyCons(h, t) => foldLeft(f, f(s, h), force t)
+    }
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `xs` going from right to left.
+    ///
+    /// That is, the result is of the form: `f(x1, ...f(xn-1, f(xn, s))...)`.
+    ///
+    pub def foldRight(f: (a, Lazy[b]) -> b, s: b, xs: LazyList[a]): b = match xs {
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/LazyList.flix:157]: I think so, something like `LazyList.foldRight((x, acc) -> x and force acc, true, LazyList.repeat(false))` does indeed work.

File: main/src/library/LazyList.flix [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        case Empty => s
+        case LazyCons(h, t) => f(h, lazy foldRight(f, s, force t))
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/LazyList.flix:159]: I think so, but because of #1546 I'm unable to test it properly.

File: main/src/library/LazyList.flix (hunk 5 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/LazyList.flix:49]: Yes. I would like these comments.

**Comment by esbenbjerre** [comment]
Do you think there should be tests for both LazyList and LazyList2? @magnus-madsen

**Comment by magnus-madsen** [comment]
> 
> 
> Do you think there should be tests for both LazyList and LazyList2? @magnus-madsen

Let us focus on LazyList and leave LazyList2 for later.

**Comment by esbenbjerre** [comment]
I think this is ready for review.
Example lazy programs:
1.
```
use LazyList.LazyList;

def isPrime(p: Int32): Bool =
    LazyList.from(2) |>
    LazyList.take(p - 2) |>
    LazyList.forall(x -> p % x != 0)

def primes(): LazyList[Int32] =
    LazyList.from(2) |>
    LazyList.filter(isPrime)
```

2. Alternative prime number sequence with sieve.
```
use LazyList.LazyList;
use LazyList.LazyList.{Empty, LazyCons};

def primes(): LazyList[Int32] = sieve(LazyList.from(2))
def sieve(ps: LazyList[Int32]): LazyList[Int32] = match ps {
    case LazyCons(p, t) => LazyCons(p, lazy sieve(LazyList.filter(x -> x % p != 0, force t)))
    case _ => Empty
}
```

3. Canonical Fibonacci sequence.
```
use LazyList.LazyList;
use LazyList.LazyList.{LazyCons};

def fibs(): LazyList[Int32] =
    LazyCons(0, lazy LazyCons(1, lazy LazyList.zipWith((x, y) -> x + y, fibs(), LazyList.tail(fibs()))))
```

File: main/src/library/LazyList.flix [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+    ///
+    pub def forall(f: a -> Bool, xs: LazyList[a]): Bool = match xs {
+        case Empty => true
+        case LazyCons(h, t) => f(h) and forall(f, force t)
+    }
+
+    ///
+    /// Returns the result of applying `f` to every element in `xs`.
+    ///
+    /// Forces the entire list `xs`.
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:213]: But mapping does not force the list!

File: main/src/library/LazyList.flix [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+    ///
+    pub def map(f: a -> b, xs: LazyList[a]): LazyList[b] = match xs {
+        case Empty => Empty
+        case LazyCons(h, t) => LazyCons(f(h), lazy map(f, force t))
+    }
+
+    ///
+    /// Returns a lazy list of every element in `xs` that satisfies the predicate `f`.
+    ///
+    /// Forces the entire list `xs`.
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:223]: Neither does filtering :)

File: main/src/library/LazyList.flix [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+        case Some((h, t)) => LazyCons(h, lazy unfold(f, t))
+    }
+
+    ///
+    /// Returns a list where the element at index `i` is `f(a, b)` where
+    /// `a` is the element at index `i` in `xs` and `b` is the element at index `i` in `ys`.
+    ///
+    /// If either `xs` or `ys` becomes depleted, then no further elements are added to the resulting list.
+    ///
+    /// Forces `Int32.min(length(xs), length(ys))` elements of `xs` and `ys`.
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:247]: I don't think this actually forces anything, since the recursive call is lazy.

File: main/src/library/LazyList.flix [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+        case _ => Empty
+    }
+
+    ///
+    /// Returns a list where the element at index `i` is `(a, b)` where
+    /// `a` is the element at index `i` in `xs` and `b` is the element at index `i` in `ys`.
+    ///
+    /// If either `xs` or `ys` becomes depleted, then no further elements are added to the resulting list.
+    ///
+    /// Forces `Int32.min(length(xs), length(ys))` elements of `xs` and `ys`.
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:260]: Same comment as above.

File: main/test/ca/uwaterloo/flix/library/TestLazyList.flix [hunk 14]
```diff
@@ -0,0 +1,519 @@
...
+
+namespace TestLazyList {
+
+    /////////////////////////////////////////////////////////////////////////////
+    // repeat                                                                  //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def repeat01(): Bool =
+        LazyList.take(1, LazyList.repeat(1)) == LazyCons(1, lazy Empty)
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/test/ca/uwaterloo/flix/library/TestLazyList.flix:12]: Hmm, so I must design the test cases to reduce to some non-lazy value?

File: main/test/ca/uwaterloo/flix/library/TestLazyList.flix [hunk 15]
```diff
@@ -0,0 +1,519 @@
...
+        let xs = LazyList.unfold(s -> if (s >= 10) None else Some(Char.fromInt32(s + 48), s + 1), 0);
+        LazyList.toList(xs) == '0' :: '1' :: '2' :: '3' :: '4' :: '5' :: '6' :: '7' :: '8' :: '9' :: Nil
+    }
+
+    /////////////////////////////////////////////////////////////////////////////
+    // fold                                                                    //
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def fold01(): Bool =
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/test/ca/uwaterloo/flix/library/TestLazyList.flix:294]: Gotcha.

File: main/src/library/LazyList.flix (hunk 10 above)
Comment (esbenbjerre) [reply main/src/library/LazyList.flix:213]: You are right. I think I misunderstood here.

File: main/src/library/LazyList.flix [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+        case Empty => ys
+        case LazyCons(h, t) => LazyCons(h, lazy append(force t, ys))
+    }
+
+    ///
+    /// Returns the reverse of `xs`.
+    ///
+    /// Forces the entire list `xs`.
+    ///
+    pub def reverse(xs: LazyList[a]): LazyList[a] = reverseHelper(xs, Empty)
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/LazyList.flix:105]: I don't think that's possible. In Haskell something like `take 1 (reverse [1..])` runs forever.

File: main/test/ca/uwaterloo/flix/library/TestLazyList.flix (hunk 14 above)
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/test/ca/uwaterloo/flix/library/TestLazyList.flix:12]: You saw that I have defined an __eq method which forces the lists, right? :)

File: main/test/ca/uwaterloo/flix/library/TestLazyList.flix [hunk 17]
```diff
@@ -0,0 +1,519 @@
...
+    def foldLeft02(): Bool =
+        LazyList.foldLeft((acc, x) -> acc + x, "", LazyList.take(3, LazyList.repeat("z"))) == "zzz"
+
+    @test
+    def foldLeft03(): Bool =
+        LazyList.foldLeft((acc, x) -> acc - x, 0, LazyList.take(3, LazyList.from(10))) == -33
+
+    @test
+    def foldLeft04(): Bool =
+        LazyList.foldLeft((acc, x) -> acc * x, 1, LazyList.take(5, LazyList.from(1))) == 120
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/test/ca/uwaterloo/flix/library/TestLazyList.flix:327]: I haven't been able to define foldLeft such that it works on infinite lists, but I don't think I'm alone:
https://www.reddit.com/r/haskell/comments/qrf54/why_does_foldr_work_on_infinite_lists_but_not/

File: main/src/library/LazyList.flix [hunk 18]
```diff
@@ -244,7 +244,7 @@ namespace LazyList {
     ///
     /// If either `xs` or `ys` becomes depleted, then no further elements are added to the resulting list.
     ///
-    /// Forces `Int32.min(length(xs), length(ys))` elements of `xs` and `ys`.
+    /// Does not force the list `xs`.
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:247]: and ys

File: main/src/library/LazyList.flix [hunk 19]
```diff
@@ -257,7 +257,7 @@ namespace LazyList {
...
```
Comment (magnus-madsen) [inline main/src/library/LazyList.flix:260]: and ys

File: main/test/ca/uwaterloo/flix/library/TestLazyList.flix (hunk 14 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/library/TestLazyList.flix:12]: > Hmm, so I must design the test cases to reduce to some non-lazy value?

Yes

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/library/TestLazyList.flix:12]: > You saw that I have defined an __eq method which forces the lists, right? :)

OK, so that works now, but in the near future we will have a type class for Eq and we will *not* allow equality on lazy values. (This default behavior would be dangerous, e.g. if you have infinite lists). 

So I agree that your solution works now, but lets future-proof it immediately :)

File: main/test/ca/uwaterloo/flix/library/TestLazyList.flix (hunk 17 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/test/ca/uwaterloo/flix/library/TestLazyList.flix:307]: OK

File: main/src/library/LazyList.flix (hunk 16 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/LazyList.flix:105]: OK

**Comment by magnus-madsen** [comment]
So I think a few more updates to the comments and updates to the tests, and then this is done :)

File: main/src/library/LazyList2.flix [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+
+    ///
+    /// Returns a list where the element at index `i` is `(a, b)` where
+    /// `a` is the element at index `i` in `xs` and `b` is the element at index `i` in `ys`.
+    ///
+    /// If either `xs` or `ys` becomes depleted, then no further elements are added to the resulting list.
+    ///
+    /// Does not force the list `xs`.
+    ///
+    pub def zip(xs: LazyList2[a], ys: LazyList2[b]): LazyList2[(a, b)] = {
```
Comment (magnus-madsen) [inline main/src/library/LazyList2.flix:253]: and ys

File: main/src/library/LazyList2.flix [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+
+    ///
+    /// Returns a list where the element at index `i` is `f(a, b)` where
+    /// `a` is the element at index `i` in `xs` and `b` is the element at index `i` in `ys`.
+    ///
+    /// If either `xs` or `ys` becomes depleted, then no further elements are added to the resulting list.
+    ///
+    /// Does not force the list `xs`.
+    ///
+    pub def zipWith(f: (a, b) -> c, xs: LazyList2[a], ys: LazyList2[b]): LazyList2[c] = {
```
Comment (magnus-madsen) [inline main/src/library/LazyList2.flix:239]: and ys

File: main/test/ca/uwaterloo/flix/library/TestLazyList.flix (hunk 14 above)
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/test/ca/uwaterloo/flix/library/TestLazyList.flix:12]: Alright, just wanted to make sure that it wouldn't work moving forward :)
I'll refactor.

**Comment by magnus-madsen** [comment]
> Example lazy programs:
> 
> 1. Infinite sequence of prime numbers.
> 
> ```
> use LazyList.LazyList;
> 
> def isPrime(p: Int32): Bool =
>     LazyList.from(2) |>
>     LazyList.take(p - 2) |>
>     LazyList.forall(x -> p % x != 0)
> 
> def primes(): LazyList[Int32] =
>     LazyList.from(2) |>
>     LazyList.filter(isPrime)
> ```

I like this example.

> 2. Alternative infinite prime number sequence with sieve.
> 
> ```
> use LazyList.LazyList;
> use LazyList.LazyList.{Empty, LazyCons};
> 
> def primes(): LazyList[Int32] = sieve(LazyList.from(2))
>
> def sieve(ps: LazyList[Int32]): LazyList[Int32] = match ps {
>     case LazyCons(p, t) => LazyCons(p, lazy sieve(LazyList.filter(x -> x % p != 0, force t)))
>     case _ => Empty
> }
> ```

I like this example too!

> 
> 3. Canonical infinite Fibonacci number sequence.
> 
> ```
> use LazyList.LazyList;
> use LazyList.LazyList.{LazyCons};
> 
> def fibs(): LazyList[Int32] =
>     LazyCons(0, lazy LazyCons(1, lazy LazyList.zipWith((x, y) -> x + y, fibs(), LazyList.tail(fibs()))))
> ```
> 

I also like this example, although its a bit complicated.

OK; so I like all of them.

I would like if we had one more example (perhaps to substitute one of the others) that was more "practical". Imagine some sees Flix and want to ask: What can I use laziness for. All of these examples are neat, but they are also a bit academic. Can we come up with a practical scenario? I guess the most boring would just be a chain of maps and filters... done lazily... Can we do better?

Meanwhile, can I ask you to add the examples to the examples folder and create a PR for the flix website? (Note that the last two examples should probably be formatted to fit into the dropdown box on the website).

> These would probably look a bit more cool with an infix operator such as `#::`.

Yes. I will think about that. It is easy to add, but I am not sure how well it plays with the research I have in mind for later.

**Comment by esbenbjerre** [comment]
> I would like if we had one more example (perhaps to substitute one of the others) that was more "practical". Imagine some sees Flix and want to ask: What can I use laziness for. All of these examples are neat, but they are also a bit academic. Can we come up with a practical scenario? I guess the most boring would just be a chain of maps and filters... done lazily... Can we do better?

Here is an example I've stolen from a Haskell discussion which might be a bit less academic (although still a bit contrived)

1. Using a lazy sorting algorithm, we can take the `k` smallest numbers from a sequence in O(n + k*log(k)) time, where `n` is the length of the sequence. In a strict setting, I guess this would always run in O(n + n*log(n)).

There are also some nice chess engines exploiting lazy evaluation, but I think they are far too long.

**Comment by magnus-madsen** [comment]
> Here is an example I've stolen from a Haskell discussion which might be a bit less academic (although still a bit contrived)
> 
>  ...

Hmm. That sounds very academic to me :)

I was looking for something the average programmer could appreciate.

**Comment by magnus-madsen** [comment]
I was thinking something like this:

```
def log(f: Lazy[String]): = ...

def main(): Unit = 
  // a complicated computation
  log(lazy "The result is ${foo}")
  // ...
```

Or something like that. Maybe that can be polished up?

**Comment by esbenbjerre** [comment]
> I was thinking something like this:
> 
> 
> 
> ```
> 
> def log(f: Lazy[String]): = ...
> 
> 
> 
> def main(): Unit = 
> 
>   // a complicated computation
> 
>   log(lazy "The result is ${foo}")
> 
>   // ...
> 
> ```
> 
> 
> 
> Or something like that. Maybe that can be polished up?

That looks doable. I'll agree that it is more approachable.

**Comment by magnus-madsen** [comment]
- Using Laziness for Infinite Streams
- Using Laziness for Logging
- Using Laziness to Compute Fibonacci

**Comment by esbenbjerre** [comment]
I've removed the use of equality on lazy constructors in the tests.

**Comment by magnus-madsen** [comment]
OK, you just need to add the three examples to `examples/` and to update `TestExamples`.

**Comment by magnus-madsen** [comment]
i.e. in this repo

**Comment by esbenbjerre** [comment]
> OK, you just need to add the three examples to `examples/` and to update `TestExamples`.

Done!

**Comment by magnus-madsen** [comment]
Excellent! We are done, except tests fail.

**Comment by magnus-madsen** [comment]
Does this fix also need to happen in the other repo?

**Comment by esbenbjerre** [comment]
> Does this fix also need to happen in the other repo?

Yes, I've opened https://github.com/flix/flix.dev/pull/18.

**Comment by magnus-madsen** [comment]
Nice!
